### ISSUE TITLE
Use JSON3 struct mapping instead of named tuples

### ISSUE THREAD

**Issue by kskyten** [laurium-labs/graphql-julia-codegen#1]
Nice project! Would it be possible to generate structs instead of named tuples and let JSON3.jl handle the serialization? So, instead of generating

```julia
const CurrentUserResult_viewer_repositories_nodes = @NamedTuple begin
  createdAt::String
  id::String
  name::AbstractString
end
```
you generate something like

```julia
using StructTypes

struct CurrentUserResult_viewer_repositories_nodes
  createdAt::String
  id::String
  name::AbstractString
end
StructTypes.StructType(::Type{CurrentUserResult_viewer_repositories_nodes}) = StructTypes.Struct()
```

**Comment by markhalonen** [comment]
I am not a Julia expert, @krrutkow can you weigh in on pros and cons of this?

**Comment by markhalonen** [comment]
I like the idea of it, seems the JSON3 package is already aimed at Code Generation: https://quinnj.github.io/JSON3.jl/dev/#Generated-Types

The original reason to use Named Tuples was that so if you query `user` in many different GraphQL queries, it is "one type" and so you only have to write one function to handle `user`.

**Comment by krrutkow** [comment]
Yes, it can certainly be done, but having separately named types with identical fields results in more cumbersome type dispatch usage.  Many different queries may request the same set of fields, and most likely Julia code would be written to dispatch on the type represented by that set of fields no matter what their query origin is.  So NamedTuple seems to be a more sensible representation of that usage, otherwise the user would need to dispatch on rather large Unions of the many equivalent types.  Using and dispatching on traits would probably be the best direction to go with it though.

**Comment by markhalonen** [comment]
> otherwise the user would need to dispatch on rather large Unions of the many equivalent types

I think a good solution to this is to use GraphQL Fragments and generate that "common type" to avoid the large type union. That's pretty much the exact reason for GraphQL Fragments.

> Using and dispatching on traits would probably be the best direction to go with it though.

Yeah, my artisan code for parsing JSON to my named tuples is not well tested in any regard.
